### PR TITLE
feat: auction engine V5

### PR DESCRIPTION
**Pull request by nieldeckx**

### CODE REVIEW

**Review by philipdesmedt — CHANGES_REQUESTED** [review]
Did a first iteration but couldn't find much interesting.

will have to do another one

File: clarity/contracts/arkadiko-liquidation-rewards-ui-v2-1.clar [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+)
+
+;; ---------------------------------------------------------
+;; Claim helpers
+;; ---------------------------------------------------------
+
+(define-public (claim-50-stx-rewards-of (reward-ids (list 50 uint)))
+  (begin
+    (map claim-stx-rewards-of reward-ids)
+    (unwrap-panic (increase-last-reward-id))
```
[the remark this answers is not in the document]
Comment (nieldeckx) [reply clarity/contracts/arkadiko-liquidation-rewards-ui-v2-1.clar]: A: the method `increase-last-reward-id` updates the `last-reward-id` of a user in steps of 25. If only called once the `last-reward-id` would be increased by max 25, while it should be more. 

**Example:**
Let's say there are 80 reward IDs in total. All for STX, except reward ID 20, 40 and 60 are DIKO rewards. 
You claim the first batch of 50 STX rewards, and the second batch of 27 STX rewards. The `last-reward-id` is still 0 as you have not yet claimed all first 25 consecutive reward IDs (reward 20 is DIKO).
Now you claim the 3 DIKO rewards in batch. The first `increase-last-reward-id` call will update `last-reward-id` to 25. Second call to 50 and third call to 75.

File: clarity/contracts/arkadiko-auction-engine-v4-2.clar [hunk 2]
```diff
@@ -0,0 +1,477 @@
+;; @contract Auction Engine - Sells off vault collateral to raise USDA
```
[the remark this answers is not in the document]
Comment (nieldeckx) [reply clarity/contracts/arkadiko-auction-engine-v4-2.clar:1]: I have updated `auction-engine-v4-1` with the new code from `auction-engine-v4-2`. And updated `arkadiko-liquidation-rewards-v1-1` with new code from `arkadiko-liquidation-rewards-v1-2`.
So you can easily check the diff.

**Comment by nieldeckx** [comment]
@philipdesmedt PR should be done and up for review.

The tests are failing because I updated old contracts with new code so you can easily diff:
https://github.com/arkadiko-dao/arkadiko/pull/501/commits/1ce2c9f5890731c8273c37cb3d974d8fb1814531

Should revert this commit later to fix tests and website.

**Comment by nieldeckx** [comment]
1) Performance improvement (DONE):
- arkadiko-liquidation-rewards-ui-v1-1
- arkadiko-liquidation-rewards-trait-v2
- arkadiko-liquidation-rewards-v1-2
- arkadiko-liquidation-rewards-ui-v1-2
- Update website

2) Governance proposal 1 (TODO):
- Use arkadiko-liquidation-rewards-v1-2
- Update script add-liquidation-pool-rewards.js on Heroku

3) Governance proposal 2 (TODO):
- Deploy & use arkadiko-auction-engine-v4-2
- Deploy & use arkadiko-stacker-payer-v3-2

File: clarity/contracts/arkadiko-liquidation-rewards-ui-v1-1.clar [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+;; ---------------------------------------------------------
+;; Claim tracking
+;; ---------------------------------------------------------
+
+(define-public (increase-last-reward-id)
+  (let (
+    (last-id (get last-reward-id (get-user-tracking tx-sender)))
+    (reward-list (map + 
+      (list u0 u1 u2 u3 u4 u5 u6 u7 u8 u9 u10 u11 u12 u13 u14 u15 u16 u17 u18 u19 u20 u21 u22 u23 u24) 
+      (list last-id last-id last-id last-id last-id last-id last-id last-id last-id last-id last-id last-id last-id last-id last-id last-id last-id last-id last-id last-id last-id last-id last-id last-id last-id)
```
Comment (philipdesmedt) [inline clarity/contracts/arkadiko-liquidation-rewards-ui-v1-1.clar:30]: Q: Might have asked this before, but do we know the costs of this and will it execute?

File: clarity/contracts/arkadiko-liquidation-rewards-ui-v1-2.clar [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+)
+
+(define-public (claim-diko-rewards-of (reward-id uint))
+  (contract-call? .arkadiko-liquidation-rewards-v1-2 claim-rewards-of reward-id .arkadiko-token .arkadiko-liquidation-pool-v1-1)
+)
+
+(define-public (claim-xbtc-rewards-of (reward-id uint))
+  ;; TODO - UPDATE ADDRESS FOR MAINNET
+  (contract-call? .arkadiko-liquidation-rewards-v1-2 claim-rewards-of reward-id .Wrapped-Bitcoin .arkadiko-liquidation-pool-v1-1)
+)
```
Comment (philipdesmedt) [inline clarity/contracts/arkadiko-liquidation-rewards-ui-v1-2.clar:100]: Q: should we add an entry for atALEX?

Comment (philipdesmedt) [reply clarity/contracts/arkadiko-liquidation-rewards-ui-v1-2.clar:100]: Also if we add a new collateral type, hwo can we add a claim method for it? We need to update the contract?

File: clarity/contracts/arkadiko-liquidation-rewards-v1-2.clar [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+
+;; ---------------------------------------------------------
+;; Add rewards
+;; ---------------------------------------------------------
+
+(define-public (add-reward (share-block uint) (token-is-stx bool) (token <ft-trait>) (total-amount uint))
+  (add-reward-locked share-block u0 token-is-stx token total-amount)
+)
+
+(define-public (add-reward-locked (share-block uint) (unlock-block uint) (token-is-stx bool) (token <ft-trait>) (total-amount uint))
```
Comment (philipdesmedt) [inline clarity/contracts/arkadiko-liquidation-rewards-v1-2.clar:114]: Q: this is usually only called from the auction engine, right?

File: clarity/contracts/arkadiko-liquidation-rewards-v1-2.clar [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+;; Init
+;; ---------------------------------------------------------
+
+(begin
+  (map-set reward-tokens  { token: .arkadiko-token } { whitelisted: true })
+  (map-set reward-tokens  { token: .xstx-token } { whitelisted: true })
+
+  ;; TODO - UPDATE ADDRESS FOR MAINNET
+  (map-set reward-tokens  { token: .Wrapped-Bitcoin } { whitelisted: true })
+)
```
Comment (philipdesmedt) [inline clarity/contracts/arkadiko-liquidation-rewards-v1-2.clar:232]: Q: add atALEX?

Comment (nieldeckx) [reply clarity/contracts/arkadiko-liquidation-rewards-v1-2.clar:232]: Q: Yes, it's done in the deployed contract:
https://explorer.stacks.co/txid/0xd4f083a5708bc6a70fbfbc8bd6b2f647f36f72d4714a194b8a1eb600d1e7c665?chain=mainnet

File: clarity/contracts/arkadiko-liquidation-rewards-v1-2.clar (hunk 5 above)
Comment (nieldeckx) [reply clarity/contracts/arkadiko-liquidation-rewards-v1-2.clar:114]: A: correct

File: clarity/contracts/arkadiko-liquidation-rewards-ui-v1-1.clar (hunk 3 above)
Comment (nieldeckx) [reply clarity/contracts/arkadiko-liquidation-rewards-ui-v1-1.clar:30]: A: I had updated the website about a week ago to use this new contract and it seemed to work fine:
https://explorer.stacks.co/txid/0xdb3c84085fbff205093709397d1680fb204b43af39bfe942cd27148bd5890b2a?chain=mainnet

**Comment by nieldeckx** [comment]
### 1) Performance improvement (DONE):
- arkadiko-liquidation-rewards-ui-v1-1
- arkadiko-liquidation-rewards-trait-v2
- arkadiko-liquidation-rewards-v1-2
- arkadiko-liquidation-rewards-ui-v1-2
- Update website

### 2) Governance proposal 1 (DONE):
- Proposal: governance-update-auction-engine-1.js
- Update script add-liquidation-pool-rewards.js on Heroku

### 3) Governance proposal 2 (TODO):
- Proposal: governance-update-auction-engine-2.js
- **Update the vaults-redeemed map in arkadiko-stacker-payer-v3-2**
- **Update stx-redeemable in arkadiko-stacker-payer-v3-2**
- **Update sip-010-trait-ft-standard.sip-010-trait address in auction-engine**
- Deploy arkadiko-auction-engine-v4-2
- Deploy arkadiko-stacker-payer-v3-2

File: clarity/contracts/arkadiko-liquidation-rewards-ui-v1-2.clar (hunk 4 above)
Comment (nieldeckx) [reply clarity/contracts/arkadiko-liquidation-rewards-ui-v1-2.clar]: I just thought of a way to get rid of the hard coded token addresses. Will solve in a new PR later :)

**Pull request merged**